cloudinit/config/cc_locale.py

```diff
@@ -6,30 +6,46 @@
 #
 # This file is part of cloud-init. See LICENSE file for license information.
 
-"""
-Locale
-------
-**Summary:** set system locale
+"""Locale: set system locale"""
 
-Configure the system locale and apply it system wide. By default use the locale
-specified by the datasource.
-
-**Internal name:** ``cc_locale``
-
-**Module frequency:** per instance
-
-**Supported distros:** all
-
-**Config keys**::
-
-    locale: <locale str>
-    locale_configfile: <path to locale config file>
-"""
+from textwrap import dedent
 
 from cloudinit import util
+from cloudinit.config.schema import get_schema_doc, validate_cloudconfig_schema
+from cloudinit.settings import PER_INSTANCE
+
+
+frequency = PER_INSTANCE
+distros = ['all']
+schema = {
+    'id': 'cc_locale',
+    'name': 'Locale',
+    'title': 'Set system locale',
+    'description': dedent("""\
+        Configure the system locale and apply it system wide. By default use
+        the locale specified by the datasource."""),
+    'distros': distros,
+    'examples': ["TODO"],
+    'frequency': frequency,
+    'type': 'object',
+    'properties': {
+        'locale': {
+            'type': 'string',
+            'description': 'TODO',
+        },
+        'locale_configfile': {
+            'type': 'string',
+            'description': 'TODO',
+        },
+    }
+}
+
+__doc__ = get_schema_doc(schema)  # Supplement python help()
 
 
 def handle(name, cfg, cloud, log, args):
+    validate_cloudconfig_schema(cfg, schema)
+
     if len(args) != 0:
```

Comment:
There are a few cloud config modules that actuall read args (which I think is only possibly sourced from /etc/cloud/cloud.cfg(.d/*)?) 

In those modules case, I think we would need to cfg.update({'locale': args[0]}) before calling validate_cloudconfig_schema.
We also might need to think about how we would have our schema validation reject such content (as some image could have provided an invalid value for that locale)

Reply:
I also think we probably need docs on this 'secret' way to pass config options to each module in /etc/cloud/cloud.cfg 
https://github.com/canonical/cloud-init/blob/master/cloudinit/stages.py#L752-L772

Reply:
I don't think we need to validate "Args" it's not in the config passed to the module; that's a runtime decision to use that instead ... and the "secret" way is incredible awkward since you 1) have to know the list of modules defined in cloud-init by default since we have no way of appending/removing items from the config module list 2) it's not documented as you say.

Reply:
We may not need to validate `args`, but we do need to be sure that our validation isn't going to _always_ fail, breaking this use case. (Given that we don't _require_ keys to be present, I think it should be fine, but I'll spend a bit of time convincing myself of this.)

Reply:
Other places that take `args` do not do anything specific, so I think we're fine to leave this as-is.

Reply:
Agreed, since providing args is an undocumented secret, and each module that uses args makes that decision in an unique way based on the number of positional args it expects.

The two paths where folks can trigger the 'args' parsing are:
 1. hijacking/overriding the list of  cloud_final_modules, cloud_config_modules, cloud_init_modules with an /etc/cloud/cloud.cfg.d/* file and providing a list for the cfg module:
-  [cc_ssh_import_id, ALWAYS, ubuntu, chad.smith]
2. calling the CLI `cloud-init single cc_ssh_import_id --freq always ubuntu chad.smith`

Since these are not documented uses of the single module, and since each module behaves uniquely related to the positional args, I think it's safe for us to avoid schema coverage for that use.